src/tribol/physics/Mortar.cpp

```diff
@@ -631,6 +641,565 @@ void ComputeSingleMortarJacobian( SurfaceContactElem& elem )
+      -0.25*x1[8] + 0.25*x1[9] + 0.25*x1[10] - 0.25*x1[11]
+   };
+   RealT de2[3] = {
+      -0.25*x1[0] - 0.25*x1[1] + 0.25*x1[2] + 0.25*x1[3],
+      -0.25*x1[4] - 0.25*x1[5] + 0.25*x1[6] + 0.25*x1[7],
+      -0.25*x1[8] - 0.25*x1[9] + 0.25*x1[10] + 0.25*x1[11]
+   };
+   RealT n[3] = {
+      de1[1]*de2[2] - de1[2]*de2[1],
+      de1[2]*de2[0] - de1[0]*de2[2],
+      de1[0]*de2[1] - de1[1]*de2[0]
+   };
+  // clang-format on
+  RealT n_mag = std::sqrt( n[0] * n[0] + n[1] * n[1] + n[2] * n[2] );
+  for ( int d{ 0 }; d < 3; ++d ) {
+    n[d] /= n_mag;
+  }
+
+  // x1t = x1 projected to plane p (def'd by x0 and n)
+  constexpr int max_coord_size = 4 * 3;
+  RealT x1t[max_coord_size];
+  for ( int i{ 0 }; i < size1; ++i ) {
+    RealT x1diff_mag = 0.0;
+    for ( int d{ 0 }; d < 3; ++d ) {
+      x1diff_mag += n[d] * ( x1[size1 * d + i] - x0[d] );
+    }
+    for ( int d{ 0 }; d < 3; ++d ) {
+      x1t[size1 * d + i] = x1[size1 * d + i] - n[d] * x1diff_mag;
+    }
+  }
+  // x2t = x2 projected to plane p
+  RealT x2t[max_coord_size];
+  for ( int i{ 0 }; i < size2; ++i ) {
+    RealT x2diff_mag = 0.0;
+    for ( int d{ 0 }; d < 3; ++d ) {
+      x2diff_mag += n[d] * ( x2[size2 * d + i] - x0[d] );
+    }
+    for ( int d{ 0 }; d < 3; ++d ) {
+      x2t[size2 * d + i] = x2[size2 * d + i] - n[d] * x2diff_mag;
+    }
+  }
+  // Tribol's clipping algorithm
+  // create a local basis; e1 is a unit vector aligned with the first edge in element 1
+  // clang-format off
+   RealT e1[3] = {
+      x1t[1] - x1t[0],
+      x1t[5] - x1t[4],
+      x1t[9] - x1t[8]
+   };
+  // clang-format on
+  RealT e1_mag = std::sqrt( e1[0] * e1[0] + e1[1] * e1[1] + e1[2] * e1[2] );
+  for ( int d{ 0 }; d < 3; ++d ) {
+    e1[d] /= e1_mag;
+  }
+  // e2 is a unit vector = n x e1
+  // clang-format off
+   RealT e2[3] = {
+      n[1]*e1[2] - n[2]*e1[1],
+      n[2]*e1[0] - n[0]*e1[2],
+      n[0]*e1[1] - n[1]*e1[0]
+   };
+  // clang-format on
+  RealT x1t_2d[4];
+  RealT y1t_2d[4];
+  PlaneTo2DCoords( x1t, x0, e1, e2, x1t_2d, y1t_2d, size1 );
+  RealT x2t_2d[4];
+  RealT y2t_2d[4];
+  PlaneTo2DCoords( x2t, x0, e1, e2, x2t_2d, y2t_2d, size2 );
+  // coordinates need to be CCW for both faces
+  RealT x2t_2d_rev[4];
+  RealT y2t_2d_rev[4];
+  for ( int i{ 0 }; i < size2; ++i ) {
+    x2t_2d_rev[i] = x2t_2d[i];
+    y2t_2d_rev[i] = y2t_2d[i];
+  }
+  ElemReverse( x2t_2d_rev, y2t_2d_rev, size2 );
+  RealT xti_2d[8];
+  RealT yti_2d[8];
+  int overlap_poly_size = 0;
+  Intersection2DPolygonEnzyme( x1t_2d, y1t_2d, size1, x2t_2d_rev, y2t_2d_rev, size2, 1.0e-8, 1.0e-8, xti_2d, yti_2d,
+                               &overlap_poly_size );
+  RealT overlap_poly_area = Area2DPolygon( xti_2d, yti_2d, overlap_poly_size );
+  if ( overlap_poly_area <= 0.0 ) {
+    return;
+  }
+
+  // Integrate mortar matrix over the polygon
+  // 1. get base triangle integration rule
+  RealT base_rule_2d[12];
+  RealT base_weights[6];
+  {
+    RealT wt1 = 0.109951743655322;
```

Comment:
Per our conversation, you didn't seem thrilled that this type of code was embedded in this routine like this :) Let's discuss alternatives. On the other hand, having a bunch of myFunctionEnzyme() type things all over the code might lead to confusion as well, so having enzyme code embedded at the method level as much as possible might be nice. Let's also address the reworking of some of the FE machinery...if it is better for enzyme, great, but it can get confusing to have two versions (perhaps with different canonical node ordering) floating around.

Reply:
Sounds good, looking forward to cleaning some of this up!